### PR TITLE
Upgrade code quality

### PR DESCRIPTION
**Pull request by cclauss**
Upgrade Linux, GitHub Actions, and Python to current versions.

### CODE REVIEW

**Comment by LouisCastricato** [comment]
Hey can you please add a description to the PR. A few sentences about what it is that you're suggesting.

**Review by LouisCastricato — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # See https://pre-commit.com/hooks.html for more hooks
 # This should be the _latest_ version of python supported by us
 default_language_version:
-  python: python3.9
+  python: python3.11
```
Comment (jon-tow) [inline .pre-commit-config.yaml:5]: Revert this to `python3.9`. We've standardized this version internally across core contributors and do not find any benefit in upgrading now.

File: .github/workflows/code_quality.yml [hunk 2]
```diff
@@ -4,10 +4,10 @@ on: [pull_request]
...
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions/setup-python@v2
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
         with:
-          python-version: 3.9
+          python-version: 3.x
```
Comment (jon-tow) [inline .github/workflows/code_quality.yml:12]: Replace the python-versioning with a matrix config to ensure code quality across the earliest supported version (`3.9`) and the latest python version (`3.11`). E.g.
```yml
    strategy:
      matrix:
        python-version: ["3.9", "3.11"]

    steps:
      - uses: actions/checkout@v3
      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}
```

**Review by jon-tow — CHANGES_REQUESTED** [review]
Thanks, @cclauss. I've left some change requests to address. The upgrade isn't required at this moment but we could merge under the conditions I've outlined.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (cclauss) [reply .pre-commit-config.yaml:5]: What’s new in Python 3.11…
> Python 3.11 is between 10-60% faster than Python 3.10. On average, we measured a 1.25x speedup on the standard benchmark suite.

File: .github/workflows/code_quality.yml (hunk 2 above)
Comment (cclauss) [reply .github/workflows/code_quality.yml]: The deprecated pre-commit action cannot handle multiple Python versions when the Python version is set in the pre-commit config.
> RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.9'

Comment (jon-tow) [reply .github/workflows/code_quality.yml]: Ah, that's annoying. I wonder if updating the pre-commit rev would fix this?

**Pull request closed without merging**